File: data/archive/types 2.py
"""
BioModuleFactory - Core Types and Models
Python port of TypeScript implementation
"""

from pydantic import BaseModel, Field
from typing import Dict, List, Optional
from enum import Enum
from datetime import datetime


class ModuleStep(str, Enum):
    """خطوات بناء الوحدة - Module building steps"""
    BIOLOGICAL_STUDY = "biological_study"
    ARCHITECTURE_DESIGN = "architecture_design"
    DEVELOPMENT = "development"
    TESTING = "testing"
    DOCUMENTATION = "documentation"


class DeliverableStatus(str, Enum):
    """حالة التسليم - Deliverable status"""
    PENDING = "pending"
    SUBMITTED = "submitted"
    APPROVED = "approved"
    REJECTED = "rejected"


class QualityGateType(str, Enum):
    """نوع بوابة الجودة - Quality gate type"""
    BLOCKING = "blocking"  # Must pass to advance
    WARNING = "warning"    # Can proceed with warnings


class ModulePhase(str, Enum):
    """مراحل HADER - HADER phases"""
    FOUNDATION = "foundation"
    KEMET_MVP = "kemet_mvp"
    CRM_AGENT = "crm_agent"
    ECOMMERCE = "ecommerce"
    INTEGRATION_LAUNCH = "integration_launch"


class Organism(str, Enum):
    """الكائنات الحية - Bio-organisms"""
    MYCELIUM = "mycelium"
    CORVID = "corvid"
    CHAMELEON = "chameleon"
    CEPHALOPOD = "cephalopod"
    ARACHNID = "arachnid"
    ANT = "ant"
    TARDIGRADE = "tardigrade"


class Deliverable(BaseModel):
    """تسليم - Deliverable"""
    id: str
    name: str
    description: str
    required: bool
    status: DeliverableStatus = DeliverableStatus.PENDING
    file_path: Optional[str] = None
    submitted_at: Optional[datetime] = None
    approved_at: Optional[datetime] = None


class QualityGate(BaseModel):
    """بوابة الجودة - Quality gate"""
    id: str
    name: str
    description: str
    gate_type: QualityGateType
    check_function: str  # Name of the check function
    error_message: str
    passed: Optional[bool] = None
    details: Optional[Dict] = None


class StepConfig(BaseModel):
    """تكوين الخطوة - Step configuration"""
    step: ModuleStep
    name: str
    duration_weeks: str
    deliverables: List[Deliverable]
    quality_gates: List[QualityGate]
    training_lessons: List[str] = []


class BioModule(BaseModel):
    """وحدة بيولوجية - Bio-inspired module"""
    id: str
    name: str
    organism: Organism
    problem_ar: str
    problem_en: str
    solution_ar: str
    solution_en: str
    phase: ModulePhase
    tech_stack: List[str]
    estimated_duration_weeks: int
    priority: int
    biological_principles: List[str]


class ModuleState(BaseModel):
    """حالة الوحدة - Module state"""
    module_id: str
    module_name: str
    current_step: ModuleStep
    completed_steps: List[ModuleStep] = []
    deliverables: List[Deliverable] = []
    started_at: datetime
    last_updated: datetime
    estimated_completion: Optional[datetime] = None


class ValidationResult(BaseModel):
    """نتيجة التحقق - Validation result"""
    passed: bool
    blocking_failures: List[QualityGate] = []
    warnings: List[QualityGate] = []
    score: float  # 0-100
    message: str


class TrainingLesson(BaseModel):
    """درس تدريبي - Training lesson"""
    id: str
    title: str
    description: str
    duration_minutes: int
    level: str  # beginner, intermediate, advanced
    content: str
    code_examples: List[Dict] = []
    exercises: List[Dict] = []
    quiz_questions: List[Dict] = []


class LessonProgress(BaseModel):
    """تقدم الدرس - Lesson progress"""
    user_id: str
    lesson_id: str
    started_at: datetime
    completed_at: Optional[datetime] = None
    exercises_completed: List[str] = []
    quiz_score: Optional[float] = None
    passed: bool = False


class ExerciseResult(BaseModel):
    """نتيجة التمرين - Exercise result"""
    exercise_id: str
    passed: bool
    score: float
    feedback: str
    submitted_code: str
    execution_time_ms: int


class QuizResult(BaseModel):
    """نتيجة الاختبار - Quiz result"""
    quiz_id: str
    total_questions: int
    correct_answers: int
    score: float
    passed: bool
    answers: List[int]
    feedback: List[str]

File: apps/haderos-web/server/bio-module-factory/types 2.ts
/**
 * BioModuleFactory - Core Types
 *
 * Transforms the 5-step module building process into a live workflow
 * with state machine, quality gates, and training integration.
 */

/**
 * The 7 Bio-Organisms that inspire HaderOS modules
 */
export enum BioOrganism {
  MYCELIUM = 'mycelium', // Fungal networks - Resource distribution
  CORVID = 'corvid', // Crow intelligence - Learning from errors
  CHAMELEON = 'chameleon', // Adaptive coloring - Market adaptation
  CEPHALOPOD = 'cephalopod', // Octopus - Distributed decision making
  ARACHNID = 'arachnid', // Spider - Fraud detection
  ANT = 'ant', // Ant colony - Logistics optimization
  TARDIGRADE = 'tardigrade', // Water bear - Business continuity
}

/**
 * The 5 HADER Project Phases
 */
export enum HADERPhase {
  FOUNDATION = 1, // Infrastructure & Team Formation
  ECOMMERCE = 2, // E-commerce Platform & Smart Inventory
  CRM_AGENT = 3, // CRM & Intelligent Agent
  KEMET_MVP = 4, // KEMET Content & HaderOS MVP
  INTEGRATION_LAUNCH = 5, // Integration, Testing & Launch
}

/**
 * The 5 Module Building Steps
 */
export enum ModuleStep {
  BIOLOGICAL_STUDY = 1, // 1-2 weeks: Study organism behavior
  ARCHITECTURE_DESIGN = 2, // 1 week: Design system architecture
  DEVELOPMENT = 3, // 2-4 weeks: Implement the module
  TESTING = 4, // 1 week: Unit, integration, performance tests
  DOCUMENTATION = 5, // 3 days: Technical & user documentation
}

/**
 * Module Step Status
 */
export enum StepStatus {
  NOT_STARTED = 'not_started',
  IN_PROGRESS = 'in_progress',
  SUBMITTED = 'submitted',
  APPROVED = 'approved',
  REJECTED = 'rejected',
}

/**
 * Deliverable for each step
 */
export interface Deliverable {
  id: string;
  name: string;
  description: string;
  required: boolean;
  fileType?: string; // e.g., "md", "pdf", "ts", "test.ts"
  submitted: boolean;
  filePath?: string;
  submittedAt?: Date;
  approvedAt?: Date;
  rejectedReason?: string;
}

/**
 * Step Configuration
 */
export interface StepConfig {
  step: ModuleStep;
  title: string;
  description: string;
  estimatedDuration: string; // e.g., "1-2 weeks"
  deliverables: Deliverable[];
  qualityGates: QualityGate[];
  trainingMaterials: TrainingMaterial[];
}

/**
 * Quality Gate - Automated checks before advancing to next step
 */
export interface QualityGate {
  id: string;
  name: string;
  description: string;
  check: (module: ModuleState) => Promise<QualityCheckResult>;
  blocking: boolean; // If true, must pass to advance
}

/**
 * Quality Check Result
 */
export interface QualityCheckResult {
  passed: boolean;
  message: string;
  details?: Record<string, any>;
}

/**
 * Training Material for each step
 */
export interface TrainingMaterial {
  id: string;
  title: string;
  type: 'article' | 'video' | 'code' | 'interactive';
  url?: string;
  content?: string;
  organism?: BioOrganism; // Related bio-organism
}

/**
 * Bio-Module Definition
 * Based on the biological organism it mimics
 */
export interface BioModule {
  // Identity
  name: string; // e.g., "mycelium"
  displayName: string; // e.g., "Mycelium Module"
  organism: BioOrganism;

  // Biological Inspiration
  behavior: string; // Natural behavior description
  adminProblem: string; // Business problem it solves
  solutionCode: string; // Code snippet example

  // Project Integration
  lifecyclePhase: HADERPhase; // Which HADER phase it belongs to
  priority: number; // 1-7 (build order)

  // Technology Stack
  technologies: string[]; // e.g., ["TypeScript", "Redis", "Kafka"]
  dependencies: string[]; // Other modules it depends on

  // References
  references: {
    papers?: string[]; // Academic papers
    books?: string[]; // Books
    videos?: string[]; // Educational videos
  };
}

/**
 * Module State - Tracks progress through the 5 steps
 */
export interface ModuleState {
  // Identity
  moduleName: string;
  organism: BioOrganism;

  // Progress
  currentStep: ModuleStep;
  stepStatuses: Record<ModuleStep, StepStatus>;
  completedSteps: ModuleStep[];

  // Deliverables
  deliverables: Record<string, Deliverable>;

  // Timeline
  startDate: Date;
  estimatedCompletionDate?: Date;
  actualCompletionDate?: Date;
  stepStartDates: Record<ModuleStep, Date>;
  stepCompletionDates: Partial<Record<ModuleStep, Date>>;

  // Team
  assignedDevelopers: string[];
  reviewer?: string;

  // Quality
  qualityChecksPassed: string[]; // IDs of passed quality gates
  qualityChecksFailed: string[]; // IDs of failed quality gates

  // Metadata
  createdAt: Date;
  updatedAt: Date;
  createdBy: string;
}

/**
 * Module Submission - When developer submits deliverables
 */
export interface ModuleSubmission {
  moduleName: string;
  step: ModuleStep;
  deliverableId: string;
  filePath: string;
  notes?: string;
  submittedBy: string;
  submittedAt: Date;
}

/**
 * Module Review - When reviewer approves/rejects
 */
export interface ModuleReview {
  moduleName: string;
  step: ModuleStep;
  deliverableId: string;
  approved: boolean;
  feedback: string;
  reviewedBy: string;
  reviewedAt: Date;
}

/**
 * Training Session - Interactive learning
 */
export interface TrainingSession {
  id: string;
  userId: string;
  organism: BioOrganism;
  materialId: string;
  startedAt: Date;
  completedAt?: Date;
  progress: number; // 0-100
  notes?: string;
}

/**
 * Module Statistics
 */
export interface ModuleStats {
  moduleName: string;
  totalDuration: number; // Days
  stepDurations: Record<ModuleStep, number>;
  deliverableCompletionRate: number; // 0-100
  qualityGatePassRate: number; // 0-100
  teamSize: number;
  linesOfCode: number;
  testCoverage: number; // 0-100
  documentationPages: number;
}

/**
 * Factory Configuration
 */
export interface FactoryConfig {
  // Quality Requirements
  minTestCoverage: number; // Default: 80%
  minDocumentationPages: number; // Default: 3
  requireCodeReview: boolean; // Default: true

  // Workflow
  autoAdvanceOnApproval: boolean; // Default: false
  allowSkipSteps: boolean; // Default: false

  // Notifications
  notifyOnSubmission: boolean; // Default: true
  notifyOnReview: boolean; // Default: true
  notifyOnQualityGateFail: boolean; // Default: true
}

/**
 * CLI Command Result
 */
export interface CommandResult {
  success: boolean;
  message: string;
  data?: any;
  errors?: string[];
}
